## PASO 0 EXTRACCION DE DATOS (df_v0)

In [ ]:
#REALIZAMOS LA EXTRACCION DE DATOS
import pandas as pd
import numpy as np
# import openpyxl

# TENEMOS TRES DOCUMENTOS EXCEL COMO FUENTE DE BASE: LOS CARGAMOSY UNIMOS EN UNO SOLO
df_parte1 = pd.read_excel('datos _oleicos_hasta_2020.xlsx',na_values = "NULL")
df_parte2 = pd.read_excel('datos_desde_2021.xlsx',na_values = "NULL")
df_parte3 = pd.read_excel('datos_sin_oleicos_hasta_2020.xlsx',na_values = "NULL")

df_v0 = df_v0.append(df_parte1,df_parte2,df_parte3)
df_v0

In [118]:
#SUSTITUIMOS 0 POR nan
df_v0.replace(0,np.nan,inplace=True)
df_v0

TipoArticuloEscandallo  Añada  Partida     Raza Alimentacion  \
0                       Jamón   2014    102.0  IBERICO      BELLOTA   
1                       Jamón   2014    102.0  IBERICO      BELLOTA   
2                      Paleta   2014    102.0  IBERICO      BELLOTA   
3                      Paleta   2014    102.0  IBERICO      BELLOTA   
4                      Paleta   2014    102.0  IBERICO      BELLOTA   
...                       ...    ...      ...      ...          ...   
158747                 Paleta   2019      NaN      NaN          NaN   
158748                 Paleta   2019      NaN      NaN          NaN   
158749                 Paleta   2019      NaN      NaN          NaN   
158750                 Paleta   2019      NaN      NaN          NaN   
158751                 Paleta   2019      NaN      NaN          NaN   

        NumeroSerieLc  AcidoOleico  AcidoPalmitico  AcidoEstearico  \
0        1.401001e+11        53.76           21.02           10.61   
1        1.401001e+11        53.76           21.02           10.61   
2        1.411001e+11        53.76           21.02           10.61   
3        1.411001e+11        53.76           21.02           10.61   
4        1.411001e+11        53.76           21.02           10.61   
...               ...          ...             ...             ...   
158747   2.360144e+11          NaN             NaN             NaN   
158748   2.360144e+11          NaN             NaN             NaN   
158749   2.360182e+11          NaN             NaN             NaN   
158750   2.360182e+11          NaN             NaN             NaN   
158751   2.360166e+11          NaN             NaN             NaN   

        AcidoLinoleico  ...  PesoNetoCanal  Nacimiento  CodigoMatadero  \
0                 8.11  ...      16351.398         NaN             1.0   
1                 8.11  ...      16351.398         NaN             1.0   
2                 8.11  ...      16351.398         NaN             1.0   
3                 8.11  ...      16351.398         NaN             1.0   
4                 8.11  ...      16351.398         NaN             1.0   
...                ...  ...            ...         ...             ...   
158747             NaN  ...            NaN         NaN             NaN   
158748             NaN  ...            NaN         NaN             NaN   
158749             NaN  ...            NaN         NaN             NaN   
158750             NaN  ...            NaN         NaN             NaN   
158751             NaN  ...            NaN         NaN             NaN   

        NroAnimales PesoEntradaEntrada  TipoPesada  PesoSalida  \
0             117.0          12.585470      Grupal       9.400   
1             117.0          12.585470      Grupal      10.400   
2             117.0           8.141026      Grupal       5.964   
3             117.0           8.141026      Grupal       3.150   
4             117.0           8.141026      Grupal       3.054   
...             ...                ...         ...         ...   
158747          NaN           0.200000      Grupal       5.934   
158748          NaN           0.200000      Grupal       6.060   
158749          NaN           0.200000      Grupal       5.934   
158750          NaN           0.200000      Grupal       6.000   
158751          NaN           0.200000      Grupal       6.460   

                           DescripcionArticulo             FechaEntrada  \
0         JAMÓN NO APTO PARA EL CONSUMO HUMANO  2014-01-13 00:00:00.000   
1             JAMON DE BELLOTA IBERICO PASSION  2014-01-13 00:00:00.000   
2               PALETA DE BELLOTA IBÉRICA D.O.  2014-01-13 00:00:00.000   
3             PALETA  DE BELLOTA IBÉRICA DESH.  2014-01-13 00:00:00.000   
4             PALETA DE BELLOTA IB. D.O. DESH.  2014-01-13 00:00:00.000   
...                                        ...                      ...   
158747  PALETA DE BELLOTA 100% IBÉRICA PASSION  2019-01-01 00:00:00.000   
158748  PALETA DE BELLOTA 100% 

In [42]:
df_v0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158752 entries, 0 to 158751
Data columns (total 23 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   TipoArticuloEscandallo  158752 non-null  object 
 1   Añada                   158752 non-null  int64  
 2   Partida                 158728 non-null  float64
 3   Raza                    158157 non-null  object 
 4   Alimentacion            158728 non-null  object 
 5   NumeroSerieLc           158750 non-null  float64
 6   AcidoOleico             39870 non-null   float64
 7   AcidoPalmitico          39870 non-null   float64
 8   AcidoEstearico          39870 non-null   float64
 9   AcidoLinoleico          39870 non-null   float64
 10  PesoVarios              133294 non-null  float64
 11  Pesovientres            132200 non-null  float64
 12  PesoCerdos              132200 non-null  float64
 13  PesoNetoCanal           132200 non-null  float64
 14  Nacimiento          

In [43]:
df_v0.nunique()

TipoArticuloEscandallo         2
Añada                          9
Partida                      524
Raza                           5
Alimentacion                   3
NumeroSerieLc             129456
AcidoOleico                  106
AcidoPalmitico               109
AcidoEstearico               103
AcidoLinoleico               106
PesoVarios                   427
Pesovientres                 422
PesoCerdos                   390
PesoNetoCanal                422
Nacimiento                   207
CodigoMatadero                 3
NroAnimales                  144
PesoEntradaEntrada          2178
TipoPesada                     2
PesoSalida                  3354
DescripcionArticulo          440
FechaEntrada                 298
FechaSalida                 1697
dtype: int64

In [1]:
df_v0.isnull()

NameError: name 'df_v0' is not defined

## PASO 1 DATA ENGINEERING: MISSING DATA
    TipoArticuloEscandallo         0
    Añada                          0
    Partida                       24 <-- 1. Registros sin partida erróneos --> Eliminar
    Raza                         595 <-- 2. Registros se clasifican por alimentación y raza --> Eliminar
    Alimentacion                  24 <-- 3. Registros se clasifican por alimentación y raza --> Eliminar
    NumeroSerieLc                  2 <-- 4. Eliminar
    AcidoOleico               118882 <-- Mantener
    AcidoPalmitico            118882 <-- Mantener
    AcidoEstearico            118882 <-- Mantener
    AcidoLinoleico            118882 <-- Mantener
    PesoVarios                 25458 <-- Mantener 
    Pesovientres               26552 <-- Mantener 
    PesoCerdos                 26552 <-- Mantener 
    PesoNetoCanal              26552 <-- Mantener 
    Nacimiento                 29730 <-- Número alto de registros, si solo falta nacimiento --> Mantener
    CodigoMatadero             16301 <-- No tiene mucha relevancia --> Mantener
    NroAnimales                 7852 <-- Hay partidas únicamente de jamones y unicamente de paletas: no relevante --> Mantener
    PesoEntradaEntrada             0
    TipoPesada                     0
    PesoSalida                 19426 <-- Registros de piezas que aún no han salido --> Mantener
    DescripcionArticulo            9 <-- No relevante --> Mantener
    FechaEntrada                   0
    FechaSalida                10114 <-- Registros de piezas que aún no han salido --> Mantener
    dtype: int64

In [45]:
df_v1 = df_v0.copy()

In [46]:
# PASO 1 MISSING DATA 
# Partida                       24 <-- 1. Registros sin partida erróneos --> Eliminar

df_v1 = df_v1.drop(df_v1[df_v1['Partida'].isna()].index)
df_v1.shape[0]

158728

In [47]:
# PASO 2 MISSING DATA 
# Raza                         595 <-- 2. Registros se clasifican por alimentación y raza --> Eliminar

df_v1 = df_v1.drop(df_v1[df_v1['Raza'].isna()].index)
df_v1.shape[0]

158157

In [48]:
# PASO 3 MISSING DATA 
# Alimentacion                  24 <-- 3. Registros se clasifican por alimentación y raza --> Eliminar

df_v1 = df_v1.drop(df_v1[df_v1['Alimentacion'].isna()].index)
df_v1.shape[0]

158157

In [49]:
# PASO 4 MISSING DATA 
# NumeroSerieLc                  2 <-- 4. Eliminar

df_v1 = df_v1.drop(df_v1[df_v1['NumeroSerieLc'].isna()].index)
df_v1.shape[0]

158155

## PASO 2 DATA ENGINEERING: WRONG DATA

    TipoArticuloEscandallo         2 
    Añada                          9
    Partida                      524
    Raza                           5 <-- 1. Hay sólo 3 razas, corregir clasificación
    Alimentacion                   3
    NumeroSerieLc             129457 <-- 2. Hay muchos duplicados --> eliminar --> no sirve como ID para tabla pieza en la BBDD.
    AcidoOleico                  107
    AcidoPalmitico               110
    AcidoEstearico               104
    AcidoLinoleico               107
    PesoVarios                   428 
    Pesovientres                 423
    PesoCerdos                   391
    PesoNetoCanal                423
    Nacimiento                   207 <-- 3. Hay entradas incorrectar no útiles: borrar información de columna incorrecta
    CodigoMatadero                 4
    NroAnimales                  145
    PesoEntradaEntrada          2178 <-- 4. Todos los pesos grupales son Null ya que entregan información falsa
    TipoPesada                     2
    PesoSalida                  3355 <-- 5. Identificar pesadas que no sean de la pieza entera (p.ej. deshuesadas)
    DescripcionArticulo          440 <-- 5. Identificar articulos deshuesados para hacer columna donde se clasifica por pesada.
    FechaEntrada                 298 <-- 6. Checkear que la diferencia entre fecha de entrada y salida sea coherente
    FechaSalida                 1697
    dtype: int64

  

    
    

In [50]:
# PASO 1 WRONG DATA
# Raza                           5 <-- 1. Hay sólo 3 razas, corregir clasificación

df_v1['Raza'].replace(to_replace={'IBERICO': '50% IBERICO', 'IBERICO PURO': '100% IBERICO'}, inplace=True)
df_v1['Raza'].unique()


array(['50% IBERICO', '75% IBERICO', '100% IBERICO'], dtype=object)

In [51]:
# PASO 2 WRONG DATA
# NumeroSerieLc             129457 <-- 2. Hay muchos duplicados --> eliminar --> no sirve como ID para tabla pieza en la BBDD.
print(df_v1.shape[0])
print(df_v1["NumeroSerieLc"].unique().shape[0])
print(df_v1["NumeroSerieLc"].nunique())

158155
128892
128892


In [52]:
# PASO 2 WRONG DATA
# NumeroSerieLc             129457 <-- 2. Hay muchos duplicados --> eliminar --> no sirve como ID para tabla pieza en la BBDD.
df_v2=df_v1.copy()
df_v2.drop_duplicates(inplace=True)
df_v2.shape[0]

150433

In [53]:
# PASO 3 WRONG DATA
# Nacimiento                   207 <-- 3. Hay entradas incorrectar no útiles: borrar información de columna incorrecta
df_v3=df_v2.copy()
df_v3 = df_v3.drop(df_v3[df_v3['Nacimiento'].str.len()>10].index)
df_v3.shape[0]

148720

In [54]:
# PASO 4 WRONG DATA
# PesoEntradaEntrada          2178 <-- 4. Hay que chequear que los pesos sean coherentes
df_v4 = df_v3.copy()

df_v4["PesoEntradaEntrada"] = np.where (df_v4['TipoPesada']=="Grupal", np.nan, df_v4["PesoEntradaEntrada"])

df_v4[df_v4['TipoPesada']=="Grupal"]


TipoArticuloEscandallo  Añada  Partida         Raza Alimentacion  \
0                       Jamón   2014    102.0  50% IBERICO      BELLOTA   
1                       Jamón   2014    102.0  50% IBERICO      BELLOTA   
2                      Paleta   2014    102.0  50% IBERICO      BELLOTA   
3                      Paleta   2014    102.0  50% IBERICO      BELLOTA   
4                      Paleta   2014    102.0  50% IBERICO      BELLOTA   
...                       ...    ...      ...          ...          ...   
158723                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158724                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158725                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158726                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158727                  Jamón   2018    999.0  75% IBERICO      BELLOTA   

        NumeroSerieLc  AcidoOleico  AcidoPalmitico  AcidoEstearico  \
0        1.401001e+11        53.76           21.02           10.61   
1        1.401001e+11        53.76           21.02           10.61   
2        1.411001e+11        53.76           21.02           10.61   
3        1.411001e+11        53.76           21.02           10.61   
4        1.411001e+11        53.76           21.02           10.61   
...               ...          ...             ...             ...   
158723   2.370306e+11          NaN             NaN             NaN   
158724   2.370306e+11          NaN             NaN             NaN   
158725   2.370306e+11          NaN             NaN             NaN   
158726   2.370306e+11          NaN             NaN             NaN   
158727   1.370306e+11          NaN             NaN             NaN   

        AcidoLinoleico  ...  PesoNetoCanal  Nacimiento  CodigoMatadero  \
0                 8.11  ...      16351.398         NaN             1.0   
1                 8.11  ...      16351.398         NaN             1.0   
2                 8.11  ...      16351.398         NaN             1.0   
3                 8.11  ...      16351.398         NaN             1.0   
4                 8.11  ...      16351.398         NaN             1.0   
...                ...  ...            ...         ...             ...   
158723             NaN  ...       9649.962  10/05/2016             1.0   
158724             NaN  ...       9649.962  10/05/2016             1.0   
158725             NaN  ...       9649.962  10/05/2016             1.0   
158726             NaN  ...       9649.962  10/05/2016             1.0   
158727             NaN  ...       9649.962  10/05/2016             1.0   

        NroAnimales PesoEntradaEntrada  TipoPesada  PesoSalida  \
0             117.0                NaN      Grupal       9.400   
1             117.0                NaN      Grupal      10.400   
2             117.0                NaN      Grupal       5.964   
3             117.0                NaN      Grupal       3.150   
4             117.0                NaN      Grupal       3.054   
...             ...                ...         ...         ...   
158723         65.0                NaN      Grupal       6.383   
158724         65.0                NaN      Grupal       3.160   
158725         65.0                NaN      Grupal       6.383   
158726         65.0                NaN      Grupal       3.160   
158727         65.0                NaN      Grupal       9.500   

                         DescripcionArticulo             FechaEntrada  \
0       JAMÓN NO APTO PARA EL CONSUMO HUMANO  2014-01-13 00:00:00.000   
1           JAMON DE BELLOTA IBERICO PASSION  2014-01-13 00:00:00.000   
2             PALETA DE BELLOTA IBÉRICA D.O.  2014-01-13 00:00:00.000   
3           PALETA  DE BELLOTA IBÉRICA DESH.  2014-01-13 00:00:00.000   
4           PALETA DE BELLOTA IB. D.O. DESH.  2014-01-13 00:00:00.000   
...                                      ...                      ...   
158723             PALETA DE BELLOTA IBÉRICA  2018-02-12 00:00:00.000   

In [55]:
import re

# PASO 5 WRONG DATA
# PesoEntradaEntrada          2178 <-- 4. Hay que chequear que los pesos sean coherentes --> crear tabla de clasificación de pesadas.

# Hay pesadas deshuesadas, en blisters y de pieza entera
# Para la BBDD haremos una clasificación previa en funcion del tipo de pesada ya que a posteriori solo se usarán las pesadas
# de pieza entera.


df_v5 = df_v4.copy()

df_v5["TipoPesadaSalida"] = np.where (df_v5['DescripcionArticulo'].str.contains("DESH"), "Pesada pieza deshuesada", "Pesada pieza entera")
df_v5["TipoPesadaSalida"] = np.where (df_v5['DescripcionArticulo'].str.contains("CAJA")|df_v5['DescripcionArticulo'].str.contains("MADERA"), "Pesada blisters para caja", "Pesada pieza entera")
df_v5["EstadoProductoSalida"] = np.where (df_v5['DescripcionArticulo'].str.contains("APT"), "NO APTO PARA CONSUMO", "APTO PARA CONSUMO")


df_v5

TipoArticuloEscandallo  Añada  Partida         Raza Alimentacion  \
0                       Jamón   2014    102.0  50% IBERICO      BELLOTA   
1                       Jamón   2014    102.0  50% IBERICO      BELLOTA   
2                      Paleta   2014    102.0  50% IBERICO      BELLOTA   
3                      Paleta   2014    102.0  50% IBERICO      BELLOTA   
4                      Paleta   2014    102.0  50% IBERICO      BELLOTA   
...                       ...    ...      ...          ...          ...   
158723                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158724                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158725                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158726                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158727                  Jamón   2018    999.0  75% IBERICO      BELLOTA   

        NumeroSerieLc  AcidoOleico  AcidoPalmitico  AcidoEstearico  \
0        1.401001e+11        53.76           21.02           10.61   
1        1.401001e+11        53.76           21.02           10.61   
2        1.411001e+11        53.76           21.02           10.61   
3        1.411001e+11        53.76           21.02           10.61   
4        1.411001e+11        53.76           21.02           10.61   
...               ...          ...             ...             ...   
158723   2.370306e+11          NaN             NaN             NaN   
158724   2.370306e+11          NaN             NaN             NaN   
158725   2.370306e+11          NaN             NaN             NaN   
158726   2.370306e+11          NaN             NaN             NaN   
158727   1.370306e+11          NaN             NaN             NaN   

        AcidoLinoleico  ...  CodigoMatadero  NroAnimales  PesoEntradaEntrada  \
0                 8.11  ...             1.0        117.0                 NaN   
1                 8.11  ...             1.0        117.0                 NaN   
2                 8.11  ...             1.0        117.0                 NaN   
3                 8.11  ...             1.0        117.0                 NaN   
4                 8.11  ...             1.0        117.0                 NaN   
...                ...  ...             ...          ...                 ...   
158723             NaN  ...             1.0         65.0                 NaN   
158724             NaN  ...             1.0         65.0                 NaN   
158725             NaN  ...             1.0         65.0                 NaN   
158726             NaN  ...             1.0         65.0                 NaN   
158727             NaN  ...             1.0         65.0                 NaN   

        TipoPesada PesoSalida                   DescripcionArticulo  \
0           Grupal      9.400  JAMÓN NO APTO PARA EL CONSUMO HUMANO   
1           Grupal     10.400      JAMON DE BELLOTA IBERICO PASSION   
2           Grupal      5.964        PALETA DE BELLOTA IBÉRICA D.O.   
3           Grupal      3.150      PALETA  DE BELLOTA IBÉRICA DESH.   
4           Grupal      3.054      PALETA DE BELLOTA IB. D.O. DESH.   
...            ...        ...                                   ...   
158723      Grupal      6.383             PALETA DE BELLOTA IBÉRICA   
158724      Grupal      3.160             PALETA DE BELLOTA IBÉRICA   
158725      Grupal      6.383      PALETA  DE BELLOTA IBÉRICA DESH.   
158726      Grupal      3.160      PALETA  DE BELLOTA IBÉRICA DESH.   
158727      Grupal      9.500              JAMON DE BELLOTA IBÉRICO   

                   FechaEntrada              FechaSalida     TipoPesadaSalida  \
0       2014-01-13 00:00:00.000  2018-04-24 00:00:00.000  Pesada pieza entera   
1       2014-01-13 00:00:00.000  2018-06-04 00:00:00.000  Pesada pieza entera   
2       2014-01-13 00:00:00.000  2016-05-06 00:00:00.000  Pesada pieza entera   
3       2014-01-13 00:00:00.000  2016-12-01 00:00:00.000  Pesada pieza entera   
4       2014-01-13 00:00:00.000  2016-08-01 00

In [60]:
# PASO 6 WRONG DATA
# FechaEntrada                 298 <-- 6. Checkear que la diferencia entre fecha de entrada y salida sea coherente.

# df_v5[(round(df_v5[FechaSalida].astype("date")-FechaEntrada].astype("date"))/30)>5]
df_v6 = df_v5.copy()


from datetime import datetime as dt

df_v6['FechaEntrada'] =  pd.to_datetime(df_v6['FechaEntrada'])
df_v6['FechaSalida'] =  pd.to_datetime(df_v6['FechaSalida'])

df_v7=df_v6.drop(df_v6[(round((df_v6["FechaSalida"]-df_v6["FechaEntrada"]).dt.days)/30)<15].index)
df_v7.shape[0]

139945

## PASO 3 DATA ENGINEERING: DATA TYPE

    <class 'pandas.core.frame.DataFrame'>
    Int64Index: 152732 entries, 0 to 158727
    Data columns (total 23 columns):
     #   Column                  Non-Null Count   Dtype  
    ---  ------                  --------------   -----  
     0   TipoArticuloEscandallo  152732 non-null  object 
     1   Añada                   152732 non-null  int64  
     2   Partida                 152732 non-null  float64 <-- Convertir a tipo int64
     3   Raza                    152732 non-null  object 
     4   Alimentacion            152732 non-null  object 
     5   NumeroSerieLc           152732 non-null  float64 <-- Convertir a tipo int64
     6   AcidoOleico             39870 non-null   float64
     7   AcidoPalmitico          39870 non-null   float64
     8   AcidoEstearico          39870 non-null   float64
     9   AcidoLinoleico          39870 non-null   float64
     10  PesoVarios              132711 non-null  float64
     11  Pesovientres            131829 non-null  float64
     12  PesoCerdos              131829 non-null  float64
     13  PesoNetoCanal           131829 non-null  float64
     14  Nacimiento              129022 non-null  object  <-- Convertir a tipo Date
     15  CodigoMatadero          139775 non-null  float64 <-- Convertir a tipo int64
     16  NroAnimales             150527 non-null  float64 <-- Convertir a tipo int64
     17  PesoEntradaEntrada      152732 non-null  float64
     18  TipoPesada              152732 non-null  object 
     19  PesoSalida              135644 non-null  float64
     20  DescripcionArticulo     152723 non-null  object 
     21  FechaEntrada            152732 non-null  object <-- Convertir a tipo Date
     22  FechaSalida             143615 non-null  object <-- Convertir a tipo Date
    dtypes: float64(14), int64(1), object(8)
    memory usage: 32.0+ MB

  

    
    

    create database factoria_jamones;
    use factoria_jamones;

    CREATE TABLE `raza` (
      `id_raza` integer auto_increment,
      `desc_raza` varchar(100),
      PRIMARY KEY (`id_raza`)
    );

    CREATE TABLE `matadero` (
      `id_matadero` integer,
      PRIMARY KEY (`id_matadero`)
    );

    CREATE TABLE `alimentacion` (
      `id_alimentacion` integer auto_increment,
      `desc_alimentacion` varchar(100),
      PRIMARY KEY (`id_alimentacion`)
    );

    CREATE TABLE `partida` (
      `id_partida` integer,
      `id_raza` integer,
      `añada` integer,
      `acido_oleico` float,
      `acido_palmitico` float,
      `acido_estearico` float,
      `acido_linoleico` float,
      `peso_varios` float,
      `peso_vientres` float,
      `peso_cerdos` float,
      `no_animales` integer,
      `nacimiento` date,
      `meses_sacrificio` integer,
      `id_matadero` integer,
      `id_alimentacion` integer,
      `fecha_entrada` date,
      PRIMARY KEY (`id_partida`),
      FOREIGN KEY (`id_raza`) REFERENCES `Raza`(`id_raza`),
      FOREIGN KEY (`id_matadero`) REFERENCES `Matadero`(`id_matadero`),
      FOREIGN KEY (`id_alimentacion`) REFERENCES `Alimentacion`(`id_alimentacion`)
    );

    CREATE TABLE `articulo_Final` (
      `id_articulo` integer auto_increment,
      `desc_articulo` varchar(100),
      PRIMARY KEY (`id_articulo`)
    );

    CREATE TABLE `tipo` (
      `id_tipo` integer auto_increment,
      `desc_tipo` varchar(100),
      PRIMARY KEY (`id_tipo`)
    );

    CREATE TABLE `pieza` (
      `id_pieza` integer auto_increment,
      `id_tipo` integer,
      `id_partida` integer,
      `peso_salida` float,
      `fecha_salida` date,
      `id_articulo` integer,
      `numero_serie_lc` integer,
      `tipo_pesada_salida` varchar(100),
      `estado_producto_salida` varchar(100),
      PRIMARY KEY (`id_pieza`),
      FOREIGN KEY (`id_articulo`) REFERENCES `Articulo_Final`(`id_articulo`),
      FOREIGN KEY (`id_partida`) REFERENCES `Partida`(`id_partida`),
      FOREIGN KEY (`id_tipo`) REFERENCES `Tipo`(`id_tipo`)
    );


In [61]:
df_v7

TipoArticuloEscandallo  Añada  Partida         Raza Alimentacion  \
0                       Jamón   2014    102.0  50% IBERICO      BELLOTA   
1                       Jamón   2014    102.0  50% IBERICO      BELLOTA   
2                      Paleta   2014    102.0  50% IBERICO      BELLOTA   
3                      Paleta   2014    102.0  50% IBERICO      BELLOTA   
4                      Paleta   2014    102.0  50% IBERICO      BELLOTA   
...                       ...    ...      ...          ...          ...   
158723                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158724                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158725                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158726                 Paleta   2018    999.0  75% IBERICO      BELLOTA   
158727                  Jamón   2018    999.0  75% IBERICO      BELLOTA   

        NumeroSerieLc  AcidoOleico  AcidoPalmitico  AcidoEstearico  \
0        1.401001e+11        53.76           21.02           10.61   
1        1.401001e+11        53.76           21.02           10.61   
2        1.411001e+11        53.76           21.02           10.61   
3        1.411001e+11        53.76           21.02           10.61   
4        1.411001e+11        53.76           21.02           10.61   
...               ...          ...             ...             ...   
158723   2.370306e+11          NaN             NaN             NaN   
158724   2.370306e+11          NaN             NaN             NaN   
158725   2.370306e+11          NaN             NaN             NaN   
158726   2.370306e+11          NaN             NaN             NaN   
158727   1.370306e+11          NaN             NaN             NaN   

        AcidoLinoleico  ...  CodigoMatadero  NroAnimales  PesoEntradaEntrada  \
0                 8.11  ...             1.0        117.0                 NaN   
1                 8.11  ...             1.0        117.0                 NaN   
2                 8.11  ...             1.0        117.0                 NaN   
3                 8.11  ...             1.0        117.0                 NaN   
4                 8.11  ...             1.0        117.0                 NaN   
...                ...  ...             ...          ...                 ...   
158723             NaN  ...             1.0         65.0                 NaN   
158724             NaN  ...             1.0         65.0                 NaN   
158725             NaN  ...             1.0         65.0                 NaN   
158726             NaN  ...             1.0         65.0                 NaN   
158727             NaN  ...             1.0         65.0                 NaN   

        TipoPesada PesoSalida                   DescripcionArticulo  \
0           Grupal      9.400  JAMÓN NO APTO PARA EL CONSUMO HUMANO   
1           Grupal     10.400      JAMON DE BELLOTA IBERICO PASSION   
2           Grupal      5.964        PALETA DE BELLOTA IBÉRICA D.O.   
3           Grupal      3.150      PALETA  DE BELLOTA IBÉRICA DESH.   
4           Grupal      3.054      PALETA DE BELLOTA IB. D.O. DESH.   
...            ...        ...                                   ...   
158723      Grupal      6.383             PALETA DE BELLOTA IBÉRICA   
158724      Grupal      3.160             PALETA DE BELLOTA IBÉRICA   
158725      Grupal      6.383      PALETA  DE BELLOTA IBÉRICA DESH.   
158726      Grupal      3.160      PALETA  DE BELLOTA IBÉRICA DESH.   
158727      Grupal      9.500              JAMON DE BELLOTA IBÉRICO   

        FechaEntrada  FechaSalida     TipoPesadaSalida  EstadoProductoSalida  
0         2014-01-13   2018-04-24  Pesada pieza entera  NO APTO PARA CONSUMO  
1         2014-01-13   2018-06-04  Pesada pieza entera     APTO PARA CONSUMO  
2         2014-01-13   2016-05-06  Pesada pieza entera     APTO PARA CONSUMO  
3         2014-01-13   2016-12-01  Pesada pieza entera     APTO PARA CONSUMO  
4         2014-01-13   2016-08-01  Pesada pieza entera  

In [76]:
df_v7["Partida"]=df_v7["Partida"].astype("int")

In [112]:
#Hay que comprobar que los siguientes valores sean únicos para cada partida

# TipoArticuloEscandallo      2
# Añada                       1 [1]
# Partida                     1 
# Raza                        1 [3]
# Alimentacion                1 [4]
# NumeroSerieLc             156
# AcidoOleico                 0 [6]
# AcidoPalmitico              0 [7]
# AcidoEstearico              0 [8]
# AcidoLinoleico              0 [9]
# PesoVarios                  1 [10]
# Pesovientres                1 [11]
# PesoCerdos                  1 [12]
# PesoNetoCanal               1 [13]
# Nacimiento                  1 [14]
# CodigoMatadero              1 [15]
# NroAnimales                 1 [16]
# PesoEntradaEntrada          0
# TipoPesada                  1 [18]
# PesoSalida                122
# DescripcionArticulo         8
# FechaEntrada                1 [21]
# FechaSalida                62
# TipoPesadaSalida            1
# EstadoProductoSalida        2
# dtype: int64
    
checkValues = [1,3,4,7,8,9,10,11,12,13,14,15,16,18,21]

for partidas in df_v7["Partida"].unique():
    for index in checkValues:
        if df_v7[df_v7["Partida"]==partidas].nunique()[index]==1 or df_v7[df_v7["Partida"]==partidas].nunique()[index]==0:
            continue
        else:
            print("Partida %d tiene repetidas en index %d" %(partidas,index))

Partida 1065 tiene repetidas en index 1
Partida 1065 tiene repetidas en index 3
Partida 1065 tiene repetidas en index 10
Partida 1065 tiene repetidas en index 11
Partida 1065 tiene repetidas en index 12
Partida 1065 tiene repetidas en index 13
Partida 1065 tiene repetidas en index 16
Partida 1065 tiene repetidas en index 21
Partida 1333 tiene repetidas en index 1
Partida 1333 tiene repetidas en index 16
Partida 1333 tiene repetidas en index 21
Partida 1584 tiene repetidas en index 1
Partida 1584 tiene repetidas en index 4
Partida 1584 tiene repetidas en index 10
Partida 1584 tiene repetidas en index 11
Partida 1584 tiene repetidas en index 12
Partida 1584 tiene repetidas en index 13
Partida 1584 tiene repetidas en index 14
Partida 1584 tiene repetidas en index 16
Partida 1584 tiene repetidas en index 21
Partida 223 tiene repetidas en index 1
Partida 223 tiene repetidas en index 3
Partida 223 tiene repetidas en index 10
Partida 223 tiene repetidas en index 11
Partida 223 tiene repetidas

Partida 1919 tiene repetidas en index 1
Partida 1919 tiene repetidas en index 10
Partida 1919 tiene repetidas en index 11
Partida 1919 tiene repetidas en index 12
Partida 1919 tiene repetidas en index 13
Partida 1919 tiene repetidas en index 14
Partida 1919 tiene repetidas en index 16
Partida 1919 tiene repetidas en index 21
Partida 1920 tiene repetidas en index 1
Partida 1920 tiene repetidas en index 3
Partida 1920 tiene repetidas en index 10
Partida 1920 tiene repetidas en index 11
Partida 1920 tiene repetidas en index 12
Partida 1920 tiene repetidas en index 13
Partida 1920 tiene repetidas en index 14
Partida 1920 tiene repetidas en index 16
Partida 1920 tiene repetidas en index 21
Partida 2020 tiene repetidas en index 1
Partida 2020 tiene repetidas en index 3
Partida 2020 tiene repetidas en index 4
Partida 2020 tiene repetidas en index 10
Partida 2020 tiene repetidas en index 11
Partida 2020 tiene repetidas en index 12
Partida 2020 tiene repetidas en index 13
Partida 2020 tiene rep

KeyboardInterrupt: 

In [115]:
df_v7[df_v7["Partida"]==1333].nunique()

TipoArticuloEscandallo      2
Añada                       2
Partida                     1
Raza                        1
Alimentacion                1
NumeroSerieLc             525
AcidoOleico                 1
AcidoPalmitico              1
AcidoEstearico              1
AcidoLinoleico              1
PesoVarios                  1
Pesovientres                1
PesoCerdos                  1
PesoNetoCanal               1
Nacimiento                  1
CodigoMatadero              1
NroAnimales                 2
PesoEntradaEntrada          0
TipoPesada                  1
PesoSalida                183
DescripcionArticulo        13
FechaEntrada                2
FechaSalida               134
TipoPesadaSalida            2
EstadoProductoSalida        2
dtype: int64

In [116]:
df_v7[df_v7["Partida"]==1333]

TipoArticuloEscandallo  Añada  Partida         Raza Alimentacion  \
4116                   Jamón   2016     1333  50% IBERICO      BELLOTA   
4117                   Jamón   2016     1333  50% IBERICO      BELLOTA   
4118                   Jamón   2016     1333  50% IBERICO      BELLOTA   
4119                   Jamón   2016     1333  50% IBERICO      BELLOTA   
4120                   Jamón   2016     1333  50% IBERICO      BELLOTA   
...                      ...    ...      ...          ...          ...   
59384                 Paleta   2019     1333  50% IBERICO      BELLOTA   
59385                 Paleta   2019     1333  50% IBERICO      BELLOTA   
59386                 Paleta   2019     1333  50% IBERICO      BELLOTA   
59387                 Paleta   2019     1333  50% IBERICO      BELLOTA   
59388                 Paleta   2019     1333  50% IBERICO      BELLOTA   

       NumeroSerieLc  AcidoOleico  AcidoPalmitico  AcidoEstearico  \
4116    1.370180e+11        55.34           18.55            7.84   
4117    1.370180e+11        55.34           18.55            7.84   
4118    1.370180e+11        55.34           18.55            7.84   
4119    1.370180e+11        55.34           18.55            7.84   
4120    1.370180e+11        55.34           18.55            7.84   
...              ...          ...             ...             ...   
59384   2.370388e+11          NaN             NaN             NaN   
59385   2.370388e+11          NaN             NaN             NaN   
59386   2.370388e+11          NaN             NaN             NaN   
59387   2.370388e+11          NaN             NaN             NaN   
59388   2.370388e+11          NaN             NaN             NaN   

       AcidoLinoleico  ...  CodigoMatadero  NroAnimales  PesoEntradaEntrada  \
4116            12.01  ...             1.0        125.0                 NaN   
4117            12.01  ...             1.0        125.0                 NaN   
4118            12.01  ...             1.0        125.0                 NaN   
4119            12.01  ...             1.0        125.0                 NaN   
4120            12.01  ...             1.0        125.0                 NaN   
...               ...  ...             ...          ...                 ...   
59384             NaN  ...             1.0         20.0                 NaN   
59385             NaN  ...             1.0         20.0                 NaN   
59386             NaN  ...             1.0         20.0                 NaN   
59387             NaN  ...             1.0         20.0                 NaN   
59388             NaN  ...             1.0         20.0                 NaN   

       TipoPesada PesoSalida                         DescripcionArticulo  \
4116       Grupal    11.2400   JAMÓN DE BELLOTA IBÉRICO 50% Raza Ibérica   
4117       Grupal     9.1100                    JAMON DE BELLOTA IBÉRICO   
4118       Grupal    11.9135                    JAMON DE BELLOTA IBÉRICO   
4119       Grupal    11.9135                    JAMON DE BELLOTA IBÉRICO   
4120       Grupal    11.9135                    JAMON DE BELLOTA IBÉRICO   
...           ...        ...                                         ...   
59384      Grupal     4.7800           PALETA DE BELLOTA IB. CAJA MADERA   
59385      Grupal     2.2760           PALETA DE BELLOTA IB. CAJA MADERA   
59386      Grupal     3.0560            PALETA  DE BELLOTA IBÉRICA DESH.   
59387      Grupal     6.2000  PALETA DE BELLOTA IBERICA 50% Raza Ibérica   
59388      Grupal     5.4350                   PALETA DE BELLOTA IBÉRICA   

       FechaEntrada  FechaSalida           TipoPesadaSalida  \
4116     2016-02-29   2021-09-27        Pesada pieza entera   
4117     2016-02-29   2020-08-24        Pesada pieza entera   
4118     2016-02-29   2019-02-07        Pesada pieza entera   
4119     2016-02-29   2019-02-07        Pesada pieza entera   
4120     2016-02-29   2019-02-07        Pesada pieza entera   
...             ...          ...                

In [105]:
for index in checkValues:
    if df_v7[df_v7["Partida"]==999].nunique()[index]==1 or df_v7[df_v7["Partida"]==999].nunique()[index]==0:
        print("ok")
    else:
        print(index)

ok
ok
ok
ok
ok
ok
ok
ok
ok
ok
ok
ok
ok
ok
ok


In [31]:
#     CREATE TABLE `partida` (
#       `id_partida` integer,
#       `id_raza` integer,
#       `añada` integer,
#       `acido_oleico` float,
#       `acido_palmitico` float,
#       `acido_estearico` float,
#       `acido_linoleico` float,
#       `peso_varios` float,
#       `peso_vientres` float,
#       `peso_cerdos` float,
#       `no_animales` integer,
#       `nacimiento` date,
#       `meses_sacrificio` integer,
#       `id_matadero` integer,
#       `id_alimentacion` integer,
#       `fecha_entrada` date,
#       PRIMARY KEY (`id_partida`),
#       FOREIGN KEY (`id_raza`) REFERENCES `Raza`(`id_raza`),
#       FOREIGN KEY (`id_matadero`) REFERENCES `Matadero`(`id_matadero`),
#       FOREIGN KEY (`id_alimentacion`) REFERENCES `Alimentacion`(`id_alimentacion`)

dict_partida = {}
for partidas in df_v7["partida"].unique:
    for index in df_v7["partida"==partidas].index:
        dict_partida[partidas]
        
        
        
        dict partida: {id_raza:[],
                       añada:[],
                       acido_oleico:[],
                       acido_palmitico:[],
                       acido_estearico:[],
                       acido_linoleico:[],
                       peso_varios:[],
                       peso_vientres:[],
                       peso_cerdos:[],
                       no_animales:[],
                       nacimiento:[],
                       meses_sacrificio:[],
                       id_matadero:[],
                       id_alimentacion:[],
                       fecha_entrada:[],}
            

SyntaxError: invalid syntax (Temp/ipykernel_7284/1654252856.py, line 29)

In [ ]:
# CREAMOS TABLA RAZA 1/7
import mysql.connector
count = 0

# CREATE TABLE `raza` (
#   `id_raza` integer auto_increment,
#   `desc_raza` varchar(100),
#   PRIMARY KEY (`id_raza`)
# );

#INTRODUCIR DATOS TABLA RAZA   
for razas in df_v_final[raza].unique:
    sqlraza = "INSERT INTO factoria_jamones.raza (desc_raza) VALUES (%s)"
    conex = mysql.connector.connect(
            host = "localhost",
            user = "root", 
            password = "",
            database ="factoria_jamones",
            port = 3306,
            autocommit=True)
    transaccion = conex.cursor()
    try:
        transaccion.execute(sqlraza, (razas))
    except: 
        print("raza número "+ str(count+1) +" ya existe en base de datos")
        count+=1

In [ ]:
# CREAMOS TABLA MATADERO 2/7
import mysql.connector
count = 0

# CREATE TABLE `matadero` (
#   `id_matadero` integer,
#   PRIMARY KEY (`id_matadero`)
# );

#INTRODUCIR DATOS TABLA MATADERO   
for mataderos in df_v_final[matadero].unique:
    sqlmatadero = "INSERT INTO factoria_jamones.matadero (id_matadero) VALUES (%s)"
    conex = mysql.connector.connect(
            host = "localhost",
            user = "root", 
            password = "",
            database ="factoria_jamones",
            port = 3306,
            autocommit=True)
    transaccion = conex.cursor()
    try:
        transaccion.execute(sqlmatadero, (mataderos))
    except: 
        print("mataderos número "+ str(count+1) +" ya existe en base de datos")
        count+=1

In [ ]:
# CREAMOS TABLA ALIMENTACIÓN 3/7
import mysql.connector
count = 0

# CREATE TABLE `alimentacion` (
#   `id_alimentacion` integer auto_increment,
#   `desc_alimentacion` varchar(100),
#   PRIMARY KEY (`id_alimentacion`)
# );

#INTRODUCIR DATOS TABLA MATADERO   
for alimentaciones in df_v_final[alimentacion].unique:
    sqlalimentacion <- "INSERT INTO `alimentacion`"= "INSERT INTO factoria_jamones.alimentacion (desc_alimentacion) VALUES (%s)"
    conex = mysql.connector.connect(
            host = "localhost",
            user = "root", 
            password = "",
            database ="factoria_jamones",
            port = 3306,
            autocommit=True)
    transaccion = conex.cursor()
    try:
        transaccion.execute(sqlalimentacion, (alimentaciones))
    except: 
        print("alimentacion número "+ str(count+1) +" ya existe en base de datos")
        count+=1

In [ ]:
def guardar_juego(juegos):
    
    import mysql.connector
    count = 0

        

            
# CREATE TABLE `matadero` (
#   `id_matadero` integer,
#   PRIMARY KEY (`id_matadero`)
# );
    
    #INTRODUCIR DATOS TABLA MATADERO   
    for mataderos in df_v_final[matadero].unique:
        sqlmatadero = "INSERT INTO factoria_jamones.matadero (id_matadero) VALUES (%s)"
        conex = mysql.connector.connect(
                host = "localhost",
                user = "root", 
                password = "",
                database ="factoria_jamones",
                port = 3306,
                autocommit=True)
        transaccion = conex.cursor()
        try:
            transaccion.execute(sqlmatadero, (mataderos))
        except: 
            print("raza número "+ str(count+1) +" ya existe en base de datos")
            count+=1
            
# CREATE TABLE `matadero` (
#   `id_matadero` integer,
#   PRIMARY KEY (`id_matadero`)
# );
    
    #INTRODUCIR DATOS TABLA MATADERO   
    for mataderos in df_v_final[matadero].unique:
        sqlmatadero = "INSERT INTO factoria_jamones.matadero (id_matadero) VALUES (%s)"
        conex = mysql.connector.connect(
                host = "localhost",
                user = "root", 
                password = "",
                database ="factoria_jamones",
                port = 3306,
                autocommit=True)
        transaccion = conex.cursor()
        try:
            transaccion.execute(sqlmatadero, (mataderos))
        except: 
            print("raza número "+ str(count+1) +" ya existe en base de datos")
            count+=1
            
        #INTRODUCIR DATOS TABLA CATEGORIAS
        for categoria in juegos[key]["recop_categorias"]:
            sqlcategorias = "INSERT INTO boardgameanalisis.categorias VALUES (%s,%s)"
            try:
                transaccion.execute(sqlcategorias,
                                    (categoria,
                                     juegos[key]["recop_categorias"][categoria]))
            except: 
                print("Categoria ya existe en BBDD")

            
        #INTRODUCIR DATOS TABLA MECANICAS
        for mecanica in juegos[key]["recop_mecanicas"]:
            sqlmecanicas = "INSERT INTO boardgameanalisis.mecanicas VALUES (%s,%s)"
            try:
                transaccion.execute(sqlmecanicas, 
                                    (mecanica,
                                    juegos[key]["recop_mecanicas"][mecanica]))
            except: 
                print("Mecanica ya existe en BBDD")

            
        #INTRODUCIR DATOS TABLA CATEGORIAS_JUEGO
        for cats in juegos[key]["recop_categorias"]:
            sqlcat_juego = "INSERT INTO boardgameanalisis.categoria_juego VALUES (%s,%s)"
            try:
                transaccion.execute(sqlcat_juego, 
                                    (int(key),int(cats)))
            except: 
                print("Relacion Juego-Categoria ya existe en BBDD")

            
            
            
        #INTRODUCIR DATOS TABLA MECANICAS_JUEGO
        for mecs in juegos[key]["recop_mecanicas"]:
            sqlmec_juego = "INSERT INTO boardgameanalisis.mecanicas_juego VALUES (%s,%s)"
            try:
                transaccion.execute(sqlmec_juego, 
                                    (int(key),int(mecs)))
            except: 
                print("Relacion Juego-Mecanica ya existe en BBDD")

In [112]:
df_v1

TipoArticuloEscandallo  Añada   Partida         Raza Alimentacion  \
0                      Paleta   2021  210247.0  75% IBERICO      BELLOTA   
1                      Paleta   2021  210247.0  75% IBERICO      BELLOTA   
2                      Paleta   2021  210247.0  75% IBERICO      BELLOTA   
3                      Paleta   2021  210247.0  75% IBERICO      BELLOTA   
4                      Paleta   2021  210247.0  75% IBERICO      BELLOTA   
...                       ...    ...       ...          ...          ...   
158747                 Paleta   2019       NaN          NaN          NaN   
158748                 Paleta   2019       NaN          NaN          NaN   
158749                 Paleta   2019       NaN          NaN          NaN   
158750                 Paleta   2019       NaN          NaN          NaN   
158751                 Paleta   2019       NaN          NaN          NaN   

        NumeroSerieLc  AcidoOleico  AcidoPalmitico  AcidoEstearico  \
0        237048812856        57.62           20.17            8.43   
1        237048812863        57.62           20.17            8.43   
2        237048812870        57.62           20.17            8.43   
3        237048812887        57.62           20.17            8.43   
4        237048812894        57.62           20.17            8.43   
...               ...          ...             ...             ...   
158747   236014424413         0.00            0.00            0.00   
158748   236014424413         0.00            0.00            0.00   
158749   236018228673         0.00            0.00            0.00   
158750   236018228673         0.00            0.00            0.00   
158751   236016641061         0.00            0.00            0.00   

        AcidoLinoleico  ...  PesoNetoCanal  Nacimiento  CodigoMatadero  \
0                  7.8  ...       18448.99  28/08/2019               1   
1                  7.8  ...       18448.99  28/08/2019               1   
2                  7.8  ...       18448.99  28/08/2019               1   
3                  7.8  ...       18448.99  28/08/2019               1   
4                  7.8  ...       18448.99  28/08/2019               1   
...                ...  ...            ...         ...             ...   
158747             0.0  ...           0.00         NaN               0   
158748             0.0  ...           0.00         NaN               0   
158749             0.0  ...           0.00         NaN               0   
158750             0.0  ...           0.00         NaN               0   
158751             0.0  ...           0.00         NaN               0   

        NroAnimales PesoEntradaEntrada  TipoPesada  PesoSalida  \
0               123               9.81  Individual       0.000   
1               123               9.66  Individual       0.000   
2               123               9.26  Individual       0.000   
3               123               9.34  Individual       0.000   
4               123               9.81  Individual       0.000   
...             ...                ...         ...         ...   
158747            0               0.20      Grupal       5.934   
158748            0               0.20      Grupal       6.060   
158749            0               0.20      Grupal       5.934   
158750            0               0.20      Grupal       6.000   
158751            0               0.20      Grupal       6.460   

                               DescripcionArticulo             FechaEntrada  \
0       PALETA DE BELLOTA IBÉRICO 75% Raza Ibérica  2021-01-18 00:00:00.000   
1       PALETA DE BELLOTA IBÉRICO 75% Raza Ibérica  2021-01-18 00:00:00.000   
2       PALETA DE BELLOTA IBÉRICO 75% Raza Ibérica  2021-01-18 00:00:00.000   
3       PALETA DE BELLOTA IBÉRICO 75% Raza Ibérica  2021-01-18 00:00:00.000   
4       PALETA DE BELLOTA IBÉRICO 75% Raza Ibérica  2021-01-18 00:00:00.000   
...                                            ...                      ...   
158747      PALETA

In [96]:
# PASO 2 ELIMINAR FECHAS DOBLES Y VACÍAS EN df_v1["Nacimiento"] 

df_v1 = df_v1.drop(df_v1[df_v1['Nacimiento'].str.len()>10].index)
df_v1 = df_v1.drop(df_v1[df_v1['Nacimiento'].isna()].index)
df_v1.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 17404 entries, 1 to 19999
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   TipoArticuloEscandallo  17404 non-null  object 
 1   Añada                   17404 non-null  int64  
 2   Partida                 17404 non-null  int64  
 3   Raza                    17404 non-null  object 
 4   Alimentacion            17404 non-null  object 
 5   NumeroSerieLc           17404 non-null  int64  
 6   AcidoOleico             17404 non-null  float64
 7   AcidoPalmitico          17404 non-null  float64
 8   AcidoEstearico          17404 non-null  float64
 9   AcidoLinoleico          17404 non-null  float64
 10  PesoVarios              17404 non-null  float64
 11  Pesovientres            17404 non-null  float64
 12  PesoCerdos              17404 non-null  float64
 13  PesoNetoCanal           17404 non-null  float64
 14  Nacimiento              17404 non-null

In [97]:
# PASO 3 Convertir columnas "Nacimiento", "FechaEntrada" y "FechaSalida" en formato fecha

df_v1['FechaEntrada'] =  pd.to_datetime(df_v1['FechaEntrada'])
df_v1['FechaSalida'] =  pd.to_datetime(df_v1['FechaSalida'])
df_v1['Nacimiento'] =  pd.to_datetime(df_v1['Nacimiento'])

In [98]:
# PASO 4 CREAMOS COLUMNA TiempoMaduracion CON FECHAS DE ENTRADA Y SALIDA DE LA PIEZA

# TiempoMaduracion EN DÍAS
df_v1["TiempoMaduracion"]= (df_v1['FechaSalida']-df_v1['FechaEntrada'])
df_v1["TiempoMaduracion"]

# TiempoMaduracionMESES EN MESES
from datetime import datetime as dt

df_v1["TiempoMaduracionMESES"]=round((df_v1["TiempoMaduracion"].dt.days)/30.41).astype(int)
df_v1["TiempoMaduracionMESES"]

1        20
2        20
3        20
4        21
5        21
         ..
19995    32
19996    21
19997    46
19998    35
19999    49
Name: TiempoMaduracionMESES, Length: 17404, dtype: int32

In [99]:
# PASO 5 CREAMOS COLUMNA EdadSacrificio CON FECHAS DE NACIMIENTO Y ENTRADA DE LA PIEZA

# EdadSacrificio EN DÍAS
df_v1['EdadSacrificio']= df_v1['FechaEntrada'] - df_v1['Nacimiento']
df_v1['EdadSacrificio']

# EdadSacrificioMESES EN MESES
df_v1["EdadSacrificioMESES"]=round((df_v1["EdadSacrificio"].dt.days)/30.41).astype(int)
df_v1["EdadSacrificioMESES"]

1        23
2        23
3        23
4        21
5        30
         ..
19995    19
19996    19
19997    15
19998    15
19999    17
Name: EdadSacrificioMESES, Length: 17404, dtype: int32